crates/daphne-server/src/router/extractor.rs

```diff
@@ -60,36 +59,76 @@ impl DecodeFromDapHttpBody for () {
     }
 }
 
+mod resource_parsers {
+    use daphne::messages::request::resource;
+    use serde::{de::DeserializeOwned, Deserialize};
+
+    #[derive(Deserialize)]
+    pub struct AggregationJobIdParser {
+        agg_job_id: resource::AggregationJobId,
+    }
+
+    #[derive(Deserialize)]
+    pub struct CollectionJobIdParser {
+        collect_job_id: resource::CollectionJobId,
+    }
+
+    #[derive(Deserialize)]
+    pub struct NoneParser;
+
+    pub trait Resource: Sized + Send + Sync {
+        type Parser: Into<Self> + Send + Sync + DeserializeOwned;
+    }
+
+    macro_rules! impl_parser {
+        ($from:ty => $to:ty |$source:pat_param| $conv:expr) => {
+            impl Resource for $to {
+                type Parser = $from;
+            }
+
+            impl ::core::convert::From<$from> for $to {
+                fn from($source: $from) -> Self {
+                    $conv
+                }
+            }
+        };
+    }
+
+    impl_parser!(AggregationJobIdParser => resource::AggregationJobId |value| value.agg_job_id);
+    impl_parser!(CollectionJobIdParser  => resource::CollectionJobId  |value| value.collect_job_id);
+    impl_parser!(NoneParser             => resource::None             |_| resource::None);
+}
+
 /// An axum extractor capable of parsing a [`DapRequest`].
 #[derive(Debug)]
-pub(super) struct UnauthenticatedDapRequestExtractor<P>(pub DapRequest<P>);
+pub(super) struct UnauthenticatedDapRequestExtractor<P, R>(pub DapRequest<P, R>);
 
 #[async_trait]
-impl<S, B, P> FromRequest<S, B, P> for UnauthenticatedDapRequestExtractor<P>
+impl<S, B, P, R> FromRequest<S, B, P> for UnauthenticatedDapRequestExtractor<P, R>
 where
     P: DecodeFromDapHttpBody,
     S: DaphneService + Send + Sync,
     B: HttpBody + Send + 'static,
     <B as HttpBody>::Data: Send,
+    R: resource_parsers::Resource,
 {
     type Rejection = AxumDapResponse;
 
     async fn from_request(req: Request<B>, state: &S) -> Result<Self, Self::Rejection> {
         #[derive(Debug, Deserialize)]
         #[serde(deny_unknown_fields)]
-        struct PathParams {
+        struct PathParams<R: resource_parsers::Resource> {
             version: DapVersion,
             task_id: TaskId,
-            agg_job_id: Option<AggregationJobId>,
-            collect_job_id: Option<CollectionJobId>,
+            #[serde(flatten)]
+            resource_parser: R::Parser,
```

Comment:
What happens if deserializtion fails? Will this be handled as an internal error or a DAP abort?

Reply:
If I'm being honest, I no longer completely follow the architecture of `daphne-server`. The ask here is to make sure that, if parsing a DAP request fails, the Helper responds as specified in the protocol: with 400 Bad Request and a Problems Detail document with the proper error.